### PR TITLE
feature/compile protobuf as scala

### PR DESCRIPTION
**Pull request by johnnyaug**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -27,8 +27,7 @@ scalacOptions ++= Seq("-release", "8", "-target:jvm-1.8")
 core / libraryDependencies ++= Seq("org.rocksdb" % "rocksdbjni" % "6.6.4",
   "commons-codec" % "commons-codec" % "1.15",
   "org.apache.spark" %% "spark-sql" % "3.0.1" % "provided",
-  "com.thesamet.scalapb" %% "sparksql-scalapb" % "0.11.0-RC1",
-  "com.thesamet.scalapb" %% "scalapb-runtime" % "0.10.4" % "protobuf",
+  "com.thesamet.scalapb" %% "scalapb-runtime" % scalapb.compiler.Version.scalapbVersion % "protobuf",
```
Comment (arielshaqed) [inline build.sbt:30]: Do we still need this?  Makes me sad (and may make compiling for both Spark 2.4.7 and Spark 3.0.1 somewhat harder, I think they may require different versions).

File: core/src/main/resources/catalog.proto [hunk 2]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 option go_package = "github.com/treevese/lakefs/catalog";
-option java_package = "io.treeverse.catalog";
+option java_package = "io.treeverse";
```
Comment (arielshaqed) [inline core/src/main/resources/catalog.proto]: This doesn't match the `go_package`; why do this?  (If anything, it should have been a deeper namespace: imagine Treeverse in 5 years, after it buys Digital Ocean, Azure, and AWS: having these at the top-level of `io.treeverse` will be somewhat embarrassing...)

File: core/src/main/scala/io/treeverse/clients/SSTableReader.scala [hunk 3]
```diff
@@ -18,12 +17,12 @@ private object local {
   }
 }
 
-class SSTableIterator[Proto <: Message](val it: SstFileReaderIterator, val messagePrototype: Proto) extends Iterator[Item[Proto]] with Closeable {
+class SSTableIterator[Proto <: GeneratedMessage](val it: SstFileReaderIterator, messagePrototype: GeneratedMessageCompanion[Proto]) extends Iterator[Item[Proto]] with Closeable {
```
Comment (arielshaqed) [inline core/src/main/scala/io/treeverse/clients/SSTableReader.scala]: It's no longer a prototype, so perhaps
```suggestion
class SSTableIterator[Proto <: GeneratedMessage](val it: SstFileReaderIterator, companion: GeneratedMessageCompanion[Proto]) extends Iterator[Item[Proto]] with Closeable {
```
:question:

File: core/src/main/scala/io/treeverse/clients/SSTableReader.scala [hunk 4]
```diff
@@ -18,12 +17,12 @@ private object local {
...
   // TODO(ariels): explicitly make it closeable, and figure out how to close it when used by
   //     Spark.
   override def hasNext: Boolean = it.isValid
 
-  override def close = it.close
+  override def close() : Unit = it.close()
```
Comment (arielshaqed) [inline core/src/main/scala/io/treeverse/clients/SSTableReader.scala:25]: Yeah, I've been reading some pretty similar style guides.  This way is better!

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Only blocking change request is _not_ to move the messages to a different package than we already use in Goland.

File: build.sbt (hunk 1 above)
Comment (johnnyaug) [reply build.sbt:30]: Unfortunately yes, this is what brings the protobuf types (e.g. google/protobuf/timestamp.proto).

Comment (arielshaqed) [reply build.sbt:30]: If only for that, we might just have to pull it into our code in order to cross-compile into Spark {2.4.7, 3.0.1}.  Not in this PR, though.

File: core/src/main/resources/catalog.proto (hunk 2 above)
Comment (arielshaqed) [reply core/src/main/resources/catalog.proto]: Understood that we still get a "committed" classname, so all is pretty much well.

**Review by arielshaqed — APPROVED** [review]
Cool, thanks!

If `-M5` is prerelease then let's use the latest available release.  If OTOH it's recommended or the only way it works, please *do* ignore :-)

**Pull request merged**